src/transformers/models/cpmant/tokenization_cpmant.py

```diff
@@ -0,0 +1,303 @@
+# coding=utf-8
+# Copyright 2022 The OpenBMB Team and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tokenization classes for CPMAnt."""
+import collections
+import os
+from importlib.util import find_spec
+from typing import Optional, Tuple
+
+from transformers import is_torch_available
+
+
+if is_torch_available():
+    import torch
+else:
+    raise RuntimeError("missing torch!")
```

Comment:
Removed.

> Thanks very much @pioliverse for iterating! I left a couple of comments, I think that some refactoring needs to be considered, after that we should be close to merge this! My main comments are:
> 
> * I think that you can wrap `CPMAntEmbedding` around a `nn.Embedding` layer even though scaling is needed. You can scale down after each call to the embedding module and make sure the input is scaled down before the `projection` call.
> * Make sure to inherit `CPMAntForCausalLM` from `CPMAntPreTrainedModel`, also make sure to follow the convention / good practices by checking what is done in OPT for instance: https://github.com/huggingface/transformers/blob/1543cee7c8c95ef47f832b1f37625ba2923c4994/src/transformers/models/opt/modeling_opt.py#L808
>    - this includes defining correctly a `lm_head` module, functions such as `get_input_embeddings`, `set_input_embeddings`, etc.
> * A lot of arguments from module's init seems to be unused, e.g. `init_std`. Try also to take the `config` object as a single argument from the init whenever possible (e.g. `CPMAntEncoder`)
> * Please make sure to follow the correct styling for docstrings (check my comments about that below)
> * If you have to initialize some weights with a specific distribution, try to initialize all the submodules weights inside `_init_weights` function from `CPMAntPreTrainedModel`
> * It's unclear to me why `forward` function is not defined in `CPMAntForCausalLM`
> * The code can be optimized here and there, I left some comments below on how you can achieve that
> * Please do not raise `RuntimeErrors` outside `if torch_is_available()`, otherwise `flax` & `tf` tests will fail
> 
> Again thanks a lot for your efforts!

Sorry for so many commits. I have made the following changes.
* added ```forward``` in ```CPMAntForCausalLM```, which is inherited from ```CPMAntPreTrainedModel ```.
* optimized some code, following the OPTModel.
* The modules initialized from Encoder's init arguments.
* added some docstrings.
* removed ```RuntimeErrors``` in tests.

Reply:
For```CPMAntEmbedding```, the model weights will be passed into ```self.weight```, which may lead to incorrect results if wrapped around ```nn.Embedding```,  maybe ```self.weight``` is good choice?